src/mlpack/methods/ann/layer/add_merge_impl.hpp

```diff
@@ -0,0 +1,210 @@
+/**
+ * @file methods/ann/layer/add_merge_impl.hpp
+ * @author Shubham Agrawal
+ *
+ * Implementation of the AddMerge class, which acts as a addition container.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_ADD_MERGE_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_ADD_MERGE_IMPL_HPP
+
+#include "add_merge.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<typename MatType>
+AddMergeType<MatType>::AddMergeType() :
+    MultiLayer<MatType>()
+{
+  // Nothing to do.
+}
+
+template<typename MatType>
+AddMergeType<MatType>::AddMergeType(const AddMergeType& other) :
+    MultiLayer<MatType>(other)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+AddMergeType<MatType>::AddMergeType(AddMergeType&& other) :
+    MultiLayer<MatType>(other)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+AddMergeType<MatType>& AddMergeType<MatType>::operator=(const AddMergeType& other)
+{
+  if (this != &other)
+  {
+    MultiLayer<MatType>::operator=(other);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+AddMergeType<MatType>& AddMergeType<MatType>::operator=(AddMergeType&& other)
+{
+  if (this != &other)
+  {
+    MultiLayer<MatType>::operator=(other);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void AddMergeType<MatType>::Forward(
+    const MatType& input, MatType& output)
+{
+  // Make sure training/testing mode is set right in each layer.
+  for (size_t i = 0; i < this->network.size(); ++i)
+    this->network[i]->Training() = this->training;
+
+  // Note that we use `output` for the last layer; layerOutputs is only used for
+  // intermediate values between layers.
+  if (this->network.size() > 1)
+  {
+    // Initialize memory for the forward pass (if needed).
+    this->InitializeForwardPassMemory(input.n_cols);
```

Comment:
Actually, I thought about it more---do we even need to initialize forward pass memory?  We could initialize just one result matrix, and have each layer output its forward pass into that matrix, and then add that to the output.  This would reduce the memory footprint of `AddMerge`.  This could also be done for the backwards pass.

It's up to you if you want to do anything here, since we are intending this implementation to be somewhat temporary anyway; we'll get it right when we do the DAG network for sure. :)